domains: Prompt user on old server invalidation.

kanishk98 opened the pull request:
Fixes #813.

On app load, we call `checkDomain()` every time we check for updates to the server icon and can detect if the new URL (possibly caused by corruption of `domain.json` or the server admin moving the address of the Zulip server) is not a legal Zulip org. This PR prompts the user to remove such an organisation and re-add the same after checking the URL. 

We also add an `override` parameter for this, so if there's an enterprise user facing this issue, they can also remove that organisation. This does allow them to remove an organisation using a hack - they could go to `domain.json` (which they can access since it's in their user folder), change the URL to say `google.com`, then have the app prompt them to remove the server. I'm not sure what we should do here, and implemented the override anyway. Let me know what you think. :)

---
<!--
Remove the fields that are not appropriate
Please include:
-->

**Screenshots?**

![image](https://user-images.githubusercontent.com/24617297/63847653-dadcba00-c9ab-11e9-91e2-2d8993d95b12.png)

**You have tested this PR on:**
  - [x] Windows
  - [ ] Linux/Ubuntu
  - [ ] macOS

kanishk98 committed:
domains: Prompt user on old server invalidation.

Fixes #813.

akashnimare commented:
@kanishk98 is it possible to not allow the server to be loaded? i.e. in screenshot I can see the google.com has been loaded but I think it should not happen?

kanishk98 commented:
Yeah, we can disallow the server from being loaded, but the most obvious way to do that is to wait till the `updateSavedServer()` method resolves. Since that would be bad UX (and would slow down users in almost all cases, where the server URLs are correct after addition), I didn't do that. 

We could choose to show an error screen (similar to our network failure screen) in the webview instead. Users won't get slowed down by that if their URL is correct, and even if Google does load, it'll be replaced fairly quickly. 
Let me know what you think.

akashnimare commented:
> We could choose to show an error screen (similar to our network failure screen) in the webview instead. Users won't get slowed down by that if their URL is correct, and even if Google does load, it'll be replaced fairly quickly.
Let me know what you think.

Sounds good 👍

kanishk98 committed:
domains: Show invalid-server.html on server fail.

If a pre-existing server fails validation, we show a new HTML page,
invalid-server.html instead of the faulty URL.

kanishk98 commented:
Updated the PR. Here's what the invalid server screen looks like:

![image](https://user-images.githubusercontent.com/24617297/63880165-d8e61b80-c9ea-11e9-8591-6535bc56dcc5.png)

kanishk98 commented:
@akashnimare whenever you find time, please also let me know what you think of overriding the `removeDomain()` function when the user clicks on Yes. I'm not sure if that's a good idea.

timabbott commented:
This PR seems somewhat risky; we don't want a transient network failure to result in an error encouraging users to disconnect the organization.  I like the non-modal screen, but I think we should have much clearer text about the range of possibilities, e.g.:

title: "Zulip cannot connect to {{domain}}"

This could be because:
* You're not on the Internet or your proxy is misconfigured
* There is no Zulip organization hosted at this URL
* This Zulip organization is temporarily unavailable
* This Zulip organization has moved or been deleted

[button: retry connection] [button: Go to settings]

kanishk98 commented:
Fair enough, I didn't think about the transient network issues. 
I'll remove the modal entirely, and also update the text on the screen. 

Thanks @timabbott! :)

akashnimare commented:
@kanishk98 can you please update the PR addressing Tim's comment?

kanishk98 commented:
@akashnimare I've included the fixes included in this PR in #723 instead as we'd discussed earlier (that fixes our reconnection logic too). 
I'll be fixing the `nodeIntegration` issue with that PR tonight, so we should be able to merge both fixes then.

akashnimare commented:
Okay, in that case, we should close this PR?

kanishk98 commented:
> in that case, we should close this PR?

Yeah, I just updated #723. Please have a look at that instead. I'll close this. :)
